add scipy.signal.upfirdn signal extension modes

grlee77 opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->

This PR complements the signal extension modes proposed in #10543.

#### What does this implement/fix?
<!--Please explain your changes.-->

This is an implementation of various signal extension modes such as `symmetric`, `periodic`, etc. for `scipy.signal.upfirdn`. Currently (prior to this PR) extension by zero padding is assumed. I have used names matching `numpy.pad` where possible. 

#### Additional information
<!--Any additional information you think is important.-->

Three additional modes (`smooth`, `antireflect` and `antisymmetric`) were included because they [are in PyWavelets](https://pywavelets.readthedocs.io/en/latest/ref/signal-extension-modes.html#signal-extension-modes) (from which this code is adapted). They can be removed if it is desired to just stick with methods available in `numpy.pad`.

I had already implemented this for another project. If this feature seems desirable for SciPy itself, I can proceed to polish it up and add tests here.

TODO
- [x] tests

larsoner commented:
@grlee77 feel free to rebase this and incorporate with what's there now that #10543 is in

grlee77 commented:
I have now rebased and added tests. For `resample_poly` the constant/cval case now uses the `upfirdn` implementation instead of doing global baseline subtraction/addition.

larsoner commented:
Is it worth it to Cythonize the linear subtraction, too? That way all modes can be in Cython, right?

rgommers commented:
@grlee77 is this still WIP?

grlee77 commented:
I think it is working as is, but let me take another look when I get a chance and I will remove the WIP label. I have not looked into Cythonizing the remaining case Eric mentioned, but if it seems straightforward I can add it.

grlee77 commented:
@larsoner : I have now added mode `line` which allowed simplifying `resample_poly` a bit.

This should be ready for review.

larsoner commented on scipy/signal/_upfirdn.py:
```diff
@@ -105,6 +117,21 @@ def upfirdn(h, x, up=1, down=1, axis=-1):
+    mode : str, optional
+        The signal extension mode to use. The set
+        ``{"constant", "symmetric", "reflect", "edge", "wrap"}`` correspond to
+        modes provided by ``numpy.pad``. ``"smooth"`` implements a smooth
```
```suggestion
        modes provided by :func:`numpy.pad`. ``"smooth"`` implements a smooth
```

grlee77 replied:
Is it even necessary to have :func:? I think SciPy has set `autolink` in `conf.py` so maybe just single backticks should work. I will test it locally and see

larsoner replied:
Just the single backticks should work, yeah

larsoner commented:
Thanks @grlee77 !
